File: tinDB/src/main/QuerySpecification/Query/QueryEdge.java
package Query;

// class for specifying a query edge.
// A node is labeled and has a source (Query.QueryNode object) and destination (Query.QueryNode object)
// all labels are stored in lowercase to prevent casing-errors!
// using uppercase labels doesn't throw an error but it'll still get stored in lowercase.
public class QueryEdge {

    public QueryNode source;
    public QueryNode target;
    public String label;

    public QueryEdge(QueryNode source, QueryNode target, String label) {
        this.source = source;
        this.target = target;
        this.label = label.toLowerCase();
    }

    public String toString() {

        String eps = "ε";
        String incoming;

        if (label.isEmpty()){
            incoming = eps;
        } else incoming = label;

        return String.format("(%s) -[%3s]-> (%s)", source.identifier, incoming, target.identifier);
    }

    public void print() {
        System.out.println(toString());
    }
}

File: src/main/kotlin/tin/model/query/QueryEdge.kt
package tin.model.query

class QueryEdge(
    val source: QueryNode,
    val target: QueryNode,
    val label: String
) {

    override fun toString(): String {
        val eps = "epsilon"
        val incoming: String = if (label.isEmpty()) {
            eps
        } else label
        return String.format("(%s) -[%3s]-> (%s)", source.identifier, incoming, target.identifier)
    }

    fun print() {
        println(this)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is QueryEdge) return false

        return this.checkForNodesEquality(other) &&
                label == other.label
    }

    /**
     * call the trimmed QueryNode.equals() method in order to prevent a circular dependency.
     */
    private fun checkForNodesEquality(other: QueryEdge): Boolean {
        return source.equalsWithoutEdges(other.source) &&
                target.equalsWithoutEdges(other.target)
    }

    override fun hashCode(): Int {
        var result = source.identifier.hashCode()
        result = 31 * result + target.identifier.hashCode()
        result = 31 * result + label.hashCode()
        return result
    }
}
